pandas/core/indexes/base.py

```diff
@@ -2939,7 +2942,11 @@ def union(self, other, sort=None):
                 # float | [u]int -> float  (the special case)
                 # <T>   | <T>    -> T
                 # <T>   | <U>    -> object
-                if not (is_integer_dtype(self.dtype) and is_integer_dtype(other.dtype)):
+                if is_float_dtype(self.dtype) and is_float_dtype(other.dtype):
+                    dtype = np.dtype("object")
```

Comment:
wouldnt we want to use the larger of the two dtypes?  (which would be `dtype`)

Reply:
Hmm, yeah, I think so. Pandas is lenient in type casting, and e.g. this casts inplicitly:

```python
>>> x1 = pd.array([1,2,3], dtype="float32")
>>> x2 = pd.array([1,2,3], dtype="float64")
>>> x1 + x2
<PandasArray>
[2.0, 4.0, 6.0]
Length: 3, dtype: float64  # implicit upcasting
```

I'll fix it (later).

Reply:
This has been fixed in #41773, so I'm marking this resolved.